File: csharp/main/LiteralDecimal.cs
using System;

namespace Deltix.Luminary
{
	public sealed class LiteralDecimal : Literal, ILiteral<Decimal?>
	{
		private readonly String _specialName;

		public LiteralDecimal(Decimal value)
			: base(LiteralKind.Decimal)
		{
			Value = value;
		}

		public LiteralDecimal(String name)
			: base(LiteralKind.Decimal)
		{
			_specialName = name;
		}

		/// <summary>
		/// The value of the literal or <c>null</c> if this is literal represents special constant like
		/// NaN, NULL, or infinities.
		/// </summary>
		public Decimal? Value { get; }

		public override String ToString()
		{
			return Value != null ? $"{Value:G29}d64" : _specialName;
		}
	}
}

File: java/src/main/java/deltix/luminary/LiteralDecimal.java
package deltix.luminary;

import deltix.dfp.Decimal;
import deltix.dfp.Decimal64Utils;

public class LiteralDecimal extends Literal {
    private final long value;

    public LiteralDecimal(@Decimal long value) {
        super(LiteralKind.DECIMAL);
        this.value = value;
    }

    @Decimal
    public long getValue() {
        return value;
    }

    @Override
    public String toString() {
        if (Decimal64Utils.isNaN(value))
            return "Decimal.NaN";
        if (Decimal64Utils.isPositiveInfinity(value))
            return "Decimal.POSITIVE_INFINITY";
        if (Decimal64Utils.isNegativeInfinity(value))
            return "Decimal.NEGATIVE_INFINITY";
        return Decimal64Utils.toString(value) + "d64";
    }
}
